### PR TITLE
Add dynare preprocessor executable only.

### PR DESCRIPTION
**Pull request by albop**
Background:

- dynare is a matlab toolbox which contains a C++ executable for preprocessing model
- there are opensource alternatives to the matlab part (including one in octave and another in-dev in python) which keep the preprocessor. This package will only contain the preprocessor and other package(s) will contain the python logic.
- the goal is ultimately to run everything in-browser using jupyter-lite as a for a port to emscripten-forge.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [ x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
    - I made a static build. 
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by albop** [comment]
@conda-forge/help-c-cpp

File: recipes/dynare_preprocessor/build.sh
Comment (xhochy) [inline recipes/dynare_preprocessor/build.sh:3]:
```suggestion
export BOOST_ROOT=$PREFIX
```

File: recipes/dynare_preprocessor/meta.yaml
Comment (xhochy) [inline recipes/dynare_preprocessor/meta.yaml:22]:
```suggestion
    - pkg-config
```

File: recipes/dynare_preprocessor/meta.yaml
Comment (xhochy) [inline recipes/dynare_preprocessor/meta.yaml:15]: Meson, bison, flex and the compiler should be in build, the rest belongs into host.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/dynare_preprocessor/build.sh
Comment (albop) [reply recipes/dynare_preprocessor/build.sh]: Thank you! I've done just that

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: Thank you @xhochy !
I followed exactly your suggestions, but made separate commits instead of using the github interface. Sorry for that. Can you mark the issues as resolved ?

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [inline recipes/dynare_preprocessor/meta.yaml:24]:
```suggestion
    - libboost-devel
```
Unless you need an old version of boost, the development package now has this name. If you need python, then list it explicitly instead of depending on the "boost" package (which depends on python).

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [inline recipes/dynare_preprocessor/meta.yaml:3]: Please place this package into a pre-release label until it has a stable release. Users of our channel should get stable non-experimental builds by default.

https://conda-forge.org/docs/maintainer/knowledge_base.html#pre-release-builds

From the downloaded artifacts:
```
project('dynare-preprocessor', 'cpp',
        version : '6-unstable',
        default_options : [ 'cpp_std=gnu++20', 'warning_level=2' ],
        meson_version : '>=0.64.0')
```

Comment (carterbox) [inline recipes/dynare_preprocessor/meta.yaml:3]: For packages without official releases, please ask the upstream project to make an official release. Otherwise, the version here may become out of sync with the upstream (canonical) version number.

If the upstream package is abandoned (it doesn't seem to be), then we have a convention that the package version should be something like "0.0.0.dev{{ build }}{{ git hash }}". I'd need to look up what we've done in the past, but the special build number + a run export prevents ABI breaks for non-versioned packages while maintaining version ordering.

File: recipes/dynare_preprocessor/run_test.sh
Comment (carterbox) [inline recipes/dynare_preprocessor/run_test.sh:2]:
```suggestion
dynare-preprocessor | grep -q "Missing model file!"
dynare-preprocessor $RECIPE_DIR/example1.mod | grep -q "Preprocessing completed."
```
You shouldn't need to list the full path of the installed binary. If it's not on the path in the test environment, then you package is broken!

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [inline recipes/dynare_preprocessor/meta.yaml:2]:
```suggestion
  name: dynare-preprocessor
```

File: recipes/dynare_preprocessor/build.sh
Comment (carterbox) [inline recipes/dynare_preprocessor/build.sh:4]: > I made a static build.

We strongly prefer that packages are dynamically linked on our channel in order to avoid the same packages/code being shipped, compiled, stored, and downloaded multiples times. If you need static linking of non-header-only packages, please provide justification.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml:3]: Yes, here I choose a very small number, to avoid any clash, but I'm happy to oblige, if there is 

There are two factors to consider here:
- last released version of upstream was 5.5 but hasn't switched to the new build system (meson). Based on this it would make sense to use number 5.6-dev .
- there is no separate release of the preprocessor which is a part of the whole distribution, so I don't know if it would follow a separate numbering.

I have no personal preference here, so I'll ask the package authors (I've been in touch with them before).

File: recipes/dynare_preprocessor/build.sh
Comment (albop) [reply recipes/dynare_preprocessor/build.sh:4]: It didn't work before I tried to add the static option ;-) On my computer it was still linking to some of my personal libraries outside of conda.
Let me try again without the static argument.

Comment (albop) [reply recipes/dynare_preprocessor/build.sh]: @carterbox : the build passes the online test.
If I try to install the artifact from CI and install locally it also work. But I get:
```
ldd $(which dynare-preprocessor)                                                                                                                                                                                        (base) 
        linux-vdso.so.1 (0x00007ffcc87ed000)
        libstdc++.so.6 => /home/pablo/micromamba/bin/../lib/libstdc++.so.6 (0x00007f662fc1d000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6630243000)
        libgcc_s.so.1 => /home/pablo/micromamba/bin/../lib/libgcc_s.so.1 (0x00007f6630228000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f6630223000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f662fa00000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f663021c000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f6630353000)
```
Is that ok ?
It is not totally clear to me how much of the shared libraries must be provided by conda, vs expected to be present on a reasonable system. Can you enlighten me here?
On the other hand the static build is really small in size ;-)

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml:3]: Thank you @carterbox for your feedback.
Just to make sure: do you make a difference between the experimental build and the experimental package ?
If I look at what upstream is doing, 5.5 is stable, but is using the old build system. Hypothetically, could a conda package eventually provide a stable release for 5.5, backporting only the new build system that is much simpler ?

File: recipes/dynare_preprocessor/meta.yaml
Comment (SylvainCorlay) [inline recipes/dynare_preprocessor/meta.yaml:24]: Move pkg-config and git-lfs to the build section.

Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: I'm moved pkg-config and removed git-lfs as it appears it is not needed anymore (we download an archive instead of cloning)

File: recipes/dynare_preprocessor/meta.yaml
Comment (svillemot) [reply recipes/dynare_preprocessor/meta.yaml:3]: Dynare upstream here.

We actually don’t publish releases of the Dynare preprocessor because it is only a component of Dynare, and so far we have never distributed it independently.

I think that using the version number of the whole Dynare could lead to problems, because you may want to update the preprocessor in Conda without waiting for a new Dynare release.

My suggestion would be to use a version number specific to Conda forge, that would incorporate the git hash. The scheme proposed by @carterbox looks good to me: `0.0.0.dev{{ build }}{{ git hash }}`. A slight variation could be `0.0.0.{{ date }}.{{ git hash }}`, which looks more explicit to me. But either is fine.

File: recipes/dynare_preprocessor/build.sh
Comment (svillemot) [reply recipes/dynare_preprocessor/build.sh]: The Dynare preprocessor is only linked against standard C and C++ libraries. I’m not familiar with Conda, but I would be surprised if those were not shipped as shared library by the package system.

File: recipes/dynare_preprocessor/meta.yaml
Comment (svillemot) [inline recipes/dynare_preprocessor/meta.yaml:29]: I would mention in the description that it’s only the preprocessor that’s packaged.

File: recipes/dynare_preprocessor/meta.yaml
Comment (svillemot) [inline recipes/dynare_preprocessor/meta.yaml:31]: I would put a more specific URL, namely https://git.dynare.org/Dynare/preprocessor

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml:29]: yep, I took that from your preprocessor archive, but of course, that makes sense

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml:3]: I'm deeply indifferent between both versions. I'll go for the latest proposal by @svillemot unless @carterbox (or @SylvainCorlay) has a strong opinion.

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: done

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml:22]: Not sure. Maybe it is needed by other parts of dynare other than the preprocessor. @svillemot ?

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: Done as suggested above.

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: @carterbox : New version is `0.0.0.dev0.2a63840eb534c3dcd1f50477d935ecc92294f00b` . All good ?

File: recipes/dynare_preprocessor/build.sh
Comment (albop) [reply recipes/dynare_preprocessor/build.sh]: Thank you @carterbox for looking into it. Now I understand well. I'll wait for @svillemot for confirmation that we shouldn't use libm or openlibm and remove the deps

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [reply recipes/dynare_preprocessor/meta.yaml:22]: I don't see anywhere in the logs that the build searches for or uses gsl. The linker logs show that sysroot libm is being used instead of openlibm, so if you want to use openlibm, you'll need to tweak the build to find the correct libm.

Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml:22]: Ok, then I'll try to remove it.

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [reply recipes/dynare_preprocessor/meta.yaml]: 0.0.0.dev{{ build }}.{{ git hash }} should be fine.

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: Actually, I tried with `libboost-headers`, and got an error so I reverted to `libboost-devel`. Is that OK?

File: recipes/dynare_preprocessor/run_test.sh
Comment (albop) [reply recipes/dynare_preprocessor/run_test.sh]: done

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: done as requested

File: recipes/dynare_preprocessor/meta.yaml
Comment (albop) [reply recipes/dynare_preprocessor/meta.yaml]: removed

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [inline recipes/dynare_preprocessor/meta.yaml:11]:
```suggestion
{% set build = 1 %}
{% set commit = 6235e6907526d468e2c35f14c38db1e308698bc9 %}

package:
  name: dynare-preprocessor
  version: 0.0.0.dev{{ build }}.{{ commit }}

source:
  url: https://git.dynare.org/Dynare/preprocessor/-/archive/{{ commit }}/preprocessor.tar.gz
  sha256: 06f76245ff5c6ac0811438cea6b817a321dc702c188b680e2a3191b54837c4ab

build:
  number: {{ build }}
  skip: true  # [win or osx]
```
The build number and the version string need to by programatically linked.

File: recipes/dynare_preprocessor/meta.yaml
Comment (carterbox) [reply recipes/dynare_preprocessor/meta.yaml:21]: Looks like you're still using libboost-devel, though.

File: recipes/dynare_preprocessor/run_test.sh
Comment (carterbox) [inline recipes/dynare_preprocessor/run_test.sh:4]: Nothing is being grepped. Doesn't that mean your test is failing?

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/dynare_preprocessor/run_test.sh
Comment (albop) [reply recipes/dynare_preprocessor/run_test.sh:4]: No. When you run the preprocessor without modem arguments, it should print that model file is missing. When the right kind of model file is supplied the output message contains "preprocessing completed". In both cases it returns an error if output is not what I'd expected.

File: recipes/dynare_preprocessor/meta.yaml
Comment (svillemot) [reply recipes/dynare_preprocessor/meta.yaml]: `gsl` and `openlibm` are not needed and can be dropped. `boost` is used, but only its headers, not the shared libraries.

**Comment by albop** [comment]
@carterbox , @svillemot : I think all points have been adressed save for the boost library. In theory it should depend only the headers. In practice, it fails if I depend only no libboost-headers (see last two commits).

**Comment by carterbox** [comment]
Please explain why the build is failing, and I may know the correct solution. You may need to increase the verbosity of your build script.

**Comment by albop** [comment]
@carterbox: I have a satisfying explanation, I think. My script was assuming that `$PREFIX/bin`  exists. I suppose it doesn' t by default and that it was created for `boost-devel`. Creating that directory (if needed) in the build script solves the issue.

**Pull request merged**
